### ISSUE TITLE
Any way to encrypt and decrypt on different devices?

### ISSUE THREAD

**Issue by kalehv** [tozny/java-aes-crypto#30]
I am encrypting string on one device. I store salt in SQLite on that device. The encrypted text is then passed to another device which obviously has it's own salt stored in it's database and that results into `MAC stored in civ does not match computed MAC.` exception.

Is there any way to make it work across multiple devices? I could be doing something fundamentally wrong here, or even something this library is not addressing but any guidance would be much appreciated.

TIA.

**Comment by SyntaxPolice** [comment]
Thanks for the issue report, @kalehv. As you are referring to salt, I assume you are using password-based key derivation?

In that case, it is permissible to pass the salt to the authorized decrypting party, along with the Ciphertext/IV/Mac. The secrecy of the salt is helpful, but not critical for the security, assuming you have a strong-enough passphrase. Does this address your question?

**Comment by kalehv** [comment]
Thanks @SyntaxPolice for reply.

So to make sure that I understand this, let me explain what I got.

Sender device -
1. Generate new salt1
2. Generate password based key using salt1 and password.
2. Encrypt data using key.
3. Send encrypted data along with salt1.

Receiver device -
1. Receive encrypted data with salt1
2. Generate key based on password (same password as device1) and salt1.
3. Decrypt data using this key.

So basically both devices only store password in their local storage.

Is that correct?

**Comment by SyntaxPolice** [comment]
Thanks. I think the numbered items are correct but I don't think it's necessary or desirable to "store password in their local storage".  Did you mean the salt?

**Comment by kalehv** [comment]
I am sorry but I am confused a little here. So If I do not store the password, how will I be able to generate key? It calls `generateKeyFromPassword(String password, String salt)` method to generate keys to encrypt right?

**Comment by SyntaxPolice** [comment]
The password-based key generation is for when you want to encrypt or decrypt something based on a password that the user types. (The password is combined w/ the salt to generate the key.) If you want to be able to type the same password on 2 devices to decrypt the same piece of text, then you can transfer and store the salt, but you shouldn't store the password.

The advantage to this approach is that you do not need to store a key. The disadvantage is that the key is only as strong as the password (plus salt), which is usually significantly less strong than a random key.

If you ultimately want to store the key, use the random key generation functionality and store that.

**Comment by kalehv** [comment]
Is storing the random key my only option here? as of now, I am making user to enter same password on all devices but if I do not store the password, I won't be able to call `generateKeyFromPassword` multiple times. The device will know how to retrieve stored salt but won't know what the password is as it is not stored. Note that user will not be entering password multiple times. I am trying something like Pushbullet does if you use that app. They ask you to enter the same password on all devices  but claim to not store it and you do not enter password every time you open the app either.

**Comment by SyntaxPolice** [comment]
I see. That's reasonable, but there are some security concerns about this approach. In particular, if you store the password in plaintext, an attacker may extract it from the device with a variety of attacks. If the user reuses that password elsewhere (which is very common!) then this could not only compromise the data you're encrypting, but also other accounts.

I don't know what pushbullet is doing. I just read their docs on end-to-end and couldn't tell what they store. Certainly they are storing something. I'm guessing it's the derived key. I'm surprised they use a fixed salt. The fixed salt + storing the derived key is a bit problematic since you can compute rainbow tables to figure out the user's password (if you have the key) or figure out the key (for weak passwords).

I'm not sure what the *best* solution is for what you're trying to do. I think the following would be a reasonable approach, but take it with a grain of salt, it's just off the top of my head.

On device 1:
1. Generate a random AES key = k1. 
1. Type the password
1. Encrypt the plain text P1 with K1 = E1 (CiphertextIvMac)
1. Generate a password-based key = K2 (plus salt)
1. Encrypt the random key K1 with K2 = EK1
1. Send the various encrypted texts to Device 2 (E1, EK1, salt)

On device 2:
- Type the password
- generate K2 with password + salt
- Decrypt key EK1 with K2 = k1
- Decrypt plain text with K1 = P1
- Store K1 for later decryptions from that device

Then to change your password:
- Type the old password and compute K2
- Type the new password and generate K3
- Decrypt EK1 with K2 (if necessary) = K1
- Encrypt K1 with K3 = EK1a
- Note that anything previously encrypted with K1 can still be decrypted

**Comment by kalehv** [comment]
Thanks @SyntaxPolice that really helps.
